Inductor: fix incorrect result of inplace unsqueeze

blzheng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94797


This pr aims to fix the incorrect result in the following test case.
```
import torch
from torch._dynamo.testing import rand_strided

@torch._dynamo.optimize("inductor")
def fn(a):
    unsqueeze_ = torch.ops.aten.unsqueeze_.default(a, 0)
    return unsqueeze_

args = [
      ((1, 1, 1, 12, 11, 3), (396, 396, 396, 33, 3, 1), torch.int64, "cpu")
       ]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]

with torch.no_grad():
    out = fn(*args)

# expected result: (396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 12, 11, 3])
print(args[0].stride(), args[0].shape) # incorrect result: (396, 396, 396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 1, 1, 12, 11, 3])
```
**Root cause**

1. [fake_tensor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#L140) is changed during [tracer.run](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/convert_frame.py#L311), then it will [pass incorrect inputs to inductor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/output_graph.py#L670).
2. example_inputs are changed during [propagate](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/mkldnn.py#L509)


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

blzheng committed:
Inductor: fix incorrect result of inplace unsqueeze

[ghstack-poisoned]

blzheng committed:
Update on "Inductor: fix incorrect result of inplace unsqueeze"



This pr aims to fix the incorrect result in the following test case.
```
torch._dynamo.optimize("inductor")
def fn(a):
    unsqueeze_ = torch.ops.aten.unsqueeze_.default(a, 0)
    return unsqueeze_

args = [
      ((1, 1, 1, 12, 11, 3), (396, 396, 396, 33, 3, 1), torch.int64, "cpu")
       ]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]

with torch.no_grad():
    out = fn(*args)

# expected result: (396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 12, 11, 3])
print(args[0].stride(), args[0].shape) # incorrect result: (396, 396, 396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 1, 1, 12, 11, 3])
```
**Root cause**

1. [fake_tensor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#L140) is changed during [tracer.run](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/convert_frame.py#L311), then it will [pass error inputs to inductor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/output_graph.py#L670).
2. example_inputs are changed during [propagate](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/mkldnn.py#L509)


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 


[ghstack-poisoned]

blzheng commented:
@pytorchbot merge

ezyang commented:
Reallocating the ShapeEnv here is not the right thing to do. I'm trying to understand the actual problem trying to be solved here...

ezyang commented:
OK, the bug trying to be fixed here is valid, but the fix here is extremely extremely wrong for dynamic shapes. Going to revert this before it causes even nastier downstream bugs to resolve.

ezyang commented:
@pytorchbot revert -c nosignal -m "bad approach, and can lead to subtle further bugs"

ezyang reviewed (changes requested):
So, tentatively, I think the correct fix for this is to fix https://github.com/pytorch/pytorch/issues/93610 (cc @eellison). The problem is we are operating on the fake tensors as we trace through the code in dynamo, but these fake tensors have to be treated immutably during processing in dynamo. This might be pretty involved though.

ezyang commented:
@blzheng do you think you would be able to attempt the proper fix?

blzheng commented:
Thanks for your explanation. I will investigate it and have a try if this issue is not urgent.

ezyang commented:
It's about as urgent as how much you want inplace metadata mutation to dynamo trace correctly

blzheng committed:
Update on "Inductor: fix incorrect result of inplace unsqueeze"



This pr aims to fix the incorrect result in the following test case.
```
torch._dynamo.optimize("inductor")
def fn(a):
    unsqueeze_ = torch.ops.aten.unsqueeze_.default(a, 0)
    return unsqueeze_

args = [
      ((1, 1, 1, 12, 11, 3), (396, 396, 396, 33, 3, 1), torch.int64, "cpu")
       ]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]

with torch.no_grad():
    out = fn(*args)

# expected result: (396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 12, 11, 3])
print(args[0].stride(), args[0].shape) # incorrect result: (396, 396, 396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 1, 1, 12, 11, 3])
```
**Root cause**

1. [fake_tensor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#L140) is changed during [tracer.run](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/convert_frame.py#L311), then it will [pass incorrect inputs to inductor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/output_graph.py#L670).
2. example_inputs are changed during [propagate](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/mkldnn.py#L509)


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 


[ghstack-poisoned]

blzheng commented:
If my understanding is correct, I think this issue can be fixed by [cloning fake tensor here](https://github.com/pytorch/pytorch/pull/94797/commits/bea44f5eca7adfd4afc5b68a7149e8babd4a886d#diff-b99ae8eb082046d6353fe4ee536a0d846e1007c7d3b21aefae468ded18fbf13eL1125-L1128).
https://github.com/pytorch/pytorch/blob/4ab1588d9919bc1a62219a5c2393e0784ddaae70/torch/_dynamo/utils.py#L1160-L1163
-->
```
    def fake_wrapper(e):
        if isinstance(e, torch.Tensor):
            assert isinstance(e, FakeTensor)
            return torch._prims_common.clone_preserve_strides(e)
        return e
```
As described in https://github.com/pytorch/pytorch/issues/93610, VariableTracker is mutated  indirectly by [running node](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/utils.py#L1187) on fake tensor. According to my understanding, the purpose of running node is to get the output value of a given node. So I think we can copy the fake tensors directly, then pass the copied fake tensors to run_node.  It will keep VariableTracker immutable and return correct results. As the result, I think it's enough to fix this issue by cloning fake tensors before run_node.
Are there any corner cases I missed?

ezyang commented:
Two subtleties:

1. You can't "just" clone, the clone has to preserve strides and storage
2. You still need to make sure you do replacements on VariableTracker if inplace metadata mutation occurred, because if someone else referred to the same TensorVariable (e.g., same id) it would need to get updated too. This is how mutation in general works with VariableTrackers, you do explicit replacements on mutation.

blzheng committed:
Update on "Inductor: fix incorrect result of inplace unsqueeze"



This pr aims to fix the incorrect result in the following test case.
```
torch._dynamo.optimize("inductor")
def fn(a):
    unsqueeze_ = torch.ops.aten.unsqueeze_.default(a, 0)
    return unsqueeze_

args = [
      ((1, 1, 1, 12, 11, 3), (396, 396, 396, 33, 3, 1), torch.int64, "cpu")
       ]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]

with torch.no_grad():
    out = fn(*args)

# expected result: (396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 12, 11, 3])
print(args[0].stride(), args[0].shape) # incorrect result: (396, 396, 396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 1, 1, 12, 11, 3])
```
**Root cause**

1. [fake_tensor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#L140) is changed during [tracer.run](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/convert_frame.py#L311), then it will [pass incorrect inputs to inductor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/output_graph.py#L670).
2. example_inputs are changed during [propagate](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/mkldnn.py#L509)


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 


[ghstack-poisoned]

ezyang commented:
holler when you want review or to discuss design

blzheng committed:
Update on "Inductor: fix incorrect result of inplace unsqueeze"



This pr aims to fix the incorrect result in the following test case.
```
torch._dynamo.optimize("inductor")
def fn(a):
    unsqueeze_ = torch.ops.aten.unsqueeze_.default(a, 0)
    return unsqueeze_

args = [
      ((1, 1, 1, 12, 11, 3), (396, 396, 396, 33, 3, 1), torch.int64, "cpu")
       ]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]

with torch.no_grad():
    out = fn(*args)

# expected result: (396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 12, 11, 3])
print(args[0].stride(), args[0].shape) # incorrect result: (396, 396, 396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 1, 1, 12, 11, 3])
```
**Root cause**

1. [fake_tensor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#L140) is changed during [tracer.run](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/convert_frame.py#L311), then it will [pass incorrect inputs to inductor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/output_graph.py#L670).
2. example_inputs are changed during [propagate](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/mkldnn.py#L509)


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 


[ghstack-poisoned]

blzheng commented:
@ezyang I added some changes according to your comments. 

> 1. You can't "just" clone, the clone has to preserve strides and storage

For this comment, `torch._prims_common.clone_preserve_strides`  is used to clone fake tensors and preserve strides. And the data_ptr is reset to preserve storage. Does this change make sense?


> 2. You still need to make sure you do replacements on VariableTracker if inplace metadata mutation occurred, because if someone else referred to the same TensorVariable (e.g., same id) it would need to get updated too. This is how mutation in general works with VariableTrackers, you do explicit replacements on mutation.

For this comment, I added a dictionary variable named `metadata_mutated_variables` in `InstructionTranslatorBase` to record the TensorVariable and its corresponding metadata mutated TensorVariable. We will add items to `metadata_mutated_variables` when a method or function is an inplace op changing metadata. 
As for TensorVariable having inplace metadata mutation, we will replace it with muated TensorVariable in `LOAD_FAST`. Is this design reasonable?

ezyang commented:
Hmm, this doesn't seem quite right, but unfortunately I'm not familiar enough with the replacement code to be able to say exactly what we should do instead. cc @jansel 

Taking a closer look

ezyang commented:
OK, so, naively, what I was expecting was for you to use tx.replace_all when you needed to replace a TensorVariable with its new mutated variant.  Haven't reviewed the other parts of PR.

jansel commented on test/inductor/test_torchinductor.py:
```diff
@@ -6992,6 +6992,75 @@ def run_node_alt(*args, **kwargs):
+                args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]
+                torch._dynamo.config.dynamic_shapes = dynamic_shapes
+                with torch.no_grad():
+                    out = fn(*args)
```
need a `dynamo.reset()` before this

jansel commented on test/inductor/test_torchinductor.py:
```diff
@@ -6992,6 +6992,75 @@ def run_node_alt(*args, **kwargs):
+                args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]
+                torch._dynamo.config.dynamic_shapes = dynamic_shapes
+                with torch.no_grad():
+                    out = fn(*args)
```
need a `dynamo.reset()` before this

jansel commented on test/inductor/test_torchinductor.py:
```diff
@@ -6992,6 +6992,75 @@ def run_node_alt(*args, **kwargs):
+                args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]
+                torch._dynamo.config.dynamic_shapes = dynamic_shapes
+                with torch.no_grad():
+                    fn(*args)
```
need a `dynamo.reset()` before this

jansel commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -40,6 +40,10 @@
 
 
+def has_same_metadata(t1, t2):
+    return t1.size == t2.size and t1.stride == t2.stride
```
not checking dtype/device?

eellison replied:
I'm not sure of those are mutable but I would check Storage, and storage offset as well

blzheng replied:
Added dtype/device and storage offset checks.

jansel commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -402,6 +406,8 @@ def call_method(
                 ),
                 **options,
             )
+            tx.metadata_mutated_variables[self] = out
```
This won't work with the checkpoint/restore logic.

You should do something similar to:
https://github.com/pytorch/pytorch/blob/428540001d9fff9c68ad2b24b8e7794800be6c77/torch/_dynamo/symbolic_convert.py#L497-L514

Though you might need to add a case where you compare `TensorVariable.proxy` instead of `.mutable_local`

blzheng replied:
I added a function named update_locals_and_stack_ex for replacements.

eellison commented on torch/_dynamo/utils.py:
```diff
@@ -1160,6 +1160,9 @@ def get_fake_value(node, tx):
     def fake_wrapper(e):
         if isinstance(e, torch.Tensor):
             assert isinstance(e, FakeTensor)
+            new_e = torch._prims_common.clone_preserve_strides(e)
```
i don't know if this works because we use `storage._cdata` to test for aliasing through FakeTensor. 
```
import torch
x = torch.rand([4])

with torch._subclasses.FakeTensorMode():
    x = torch.rand([4])
    z = x.clone()

    x.data_ptr = z.data_ptr
    assert x.untyped_storage()._cdata == z.untyped_storage()._cdata
```

eellison replied:
Actually, all of the data ptrs of fake tensor show 0, so the above is definitely not enough to preserve aliasing

blzheng replied:
Added a function named clone_preserve_strides_storage for cloning while preserving strides and storage.

eellison commented on torch/_dynamo/variables/torch.py:
```diff
@@ -590,6 +590,14 @@ def check_any_unspec(x):
             )
+            is_inplace_func = False
+            try:
+                if fn_.__name__[-1] == "_" or fn_._name[-1] == "_":
```
why do we need a try catch here ?

blzheng replied:
For fn_ that is an instance of torch._ops.OpOverload, fn_.__name__ looks like "xxx.default". To get its opname, we use fn_._name here, but this function is not an attribute of ''builtin_function_or_method' object.

jgong5 replied:
So just check the existence of `_name` attribute before checking its value would be good enough?

eellison commented on torch/_inductor/mkldnn.py:
```diff
@@ -525,7 +525,16 @@ def mkldnn_fuse_fx(gm: torch.fx.GraphModule, example_inputs):
     fake_mode = fake_mode_from_tensors(example_inputs)
-    ShapeProp(gm, fake_mode=fake_mode).propagate(*example_inputs)
+    # clone inputs to avoid side effects caused by inplace ops during propagate
+    tmp_example_inputs = list(
```
it might not matter in mlkdnn but we should clone preserving storage aliasing/other properties. maybe through a util on FakeTensor

blzheng committed:
Update on "Inductor: fix incorrect result of inplace unsqueeze"



This pr aims to fix the incorrect result in the following test case.
```
torch._dynamo.optimize("inductor")
def fn(a):
    unsqueeze_ = torch.ops.aten.unsqueeze_.default(a, 0)
    return unsqueeze_

args = [
      ((1, 1, 1, 12, 11, 3), (396, 396, 396, 33, 3, 1), torch.int64, "cpu")
       ]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]

with torch.no_grad():
    out = fn(*args)

# expected result: (396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 12, 11, 3])
print(args[0].stride(), args[0].shape) # incorrect result: (396, 396, 396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 1, 1, 12, 11, 3])
```
**Root cause**

1. [fake_tensor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#L140) is changed during [tracer.run](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/convert_frame.py#L311), then it will [pass incorrect inputs to inductor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/output_graph.py#L670).
2. example_inputs are changed during [propagate](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/mkldnn.py#L509)


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 


[ghstack-poisoned]

jgong5 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -513,6 +513,20 @@ def skip(v: VariableTracker):
                 repl, x, cache, skip_fn=skip
             )
 
+    def update_locals_and_stack_ex(
```
I'm wondering: can we add a test case that can cover this local/stack variable update?

blzheng committed:
Update on "Inductor: fix incorrect result of inplace unsqueeze"



This pr aims to fix the incorrect result in the following test case.
```
torch._dynamo.optimize("inductor")
def fn(a):
    unsqueeze_ = torch.ops.aten.unsqueeze_.default(a, 0)
    return unsqueeze_

args = [
      ((1, 1, 1, 12, 11, 3), (396, 396, 396, 33, 3, 1), torch.int64, "cpu")
       ]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]

with torch.no_grad():
    out = fn(*args)

# expected result: (396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 12, 11, 3])
print(args[0].stride(), args[0].shape) # incorrect result: (396, 396, 396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 1, 1, 12, 11, 3])
```
**Root cause**

1. [fake_tensor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#L140) is changed during [tracer.run](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/convert_frame.py#L311), then it will [pass incorrect inputs to inductor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/output_graph.py#L670).
2. example_inputs are changed during [propagate](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/mkldnn.py#L509)


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 


[ghstack-poisoned]

blzheng committed:
Update on "Inductor: fix incorrect result of inplace unsqueeze"



This pr aims to fix the incorrect result in the following test case.
```
torch._dynamo.optimize("inductor")
def fn(a):
    unsqueeze_ = torch.ops.aten.unsqueeze_.default(a, 0)
    return unsqueeze_

args = [
      ((1, 1, 1, 12, 11, 3), (396, 396, 396, 33, 3, 1), torch.int64, "cpu")
       ]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]

with torch.no_grad():
    out = fn(*args)

# expected result: (396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 12, 11, 3])
print(args[0].stride(), args[0].shape) # incorrect result: (396, 396, 396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 1, 1, 12, 11, 3])
```
**Root cause**

1. [fake_tensor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#L140) is changed during [tracer.run](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/convert_frame.py#L311), then it will [pass incorrect inputs to inductor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/output_graph.py#L670).
2. example_inputs are changed during [propagate](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/mkldnn.py#L509)


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 


[ghstack-poisoned]

blzheng committed:
Update on "Inductor: fix incorrect result of inplace unsqueeze"



This pr aims to fix the incorrect result in the following test case.
```
torch._dynamo.optimize("inductor")
def fn(a):
    unsqueeze_ = torch.ops.aten.unsqueeze_.default(a, 0)
    return unsqueeze_

args = [
      ((1, 1, 1, 12, 11, 3), (396, 396, 396, 33, 3, 1), torch.int64, "cpu")
       ]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]

with torch.no_grad():
    out = fn(*args)

# expected result: (396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 12, 11, 3])
print(args[0].stride(), args[0].shape) # incorrect result: (396, 396, 396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 1, 1, 12, 11, 3])
```
**Root cause**

1. [fake_tensor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#L140) is changed during [tracer.run](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/convert_frame.py#L311), then it will [pass incorrect inputs to inductor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/output_graph.py#L670).
2. example_inputs are changed during [propagate](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/mkldnn.py#L509)


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 


[ghstack-poisoned]

blzheng committed:
Update on "Inductor: fix incorrect result of inplace unsqueeze"



This pr aims to fix the incorrect result in the following test case.
```
torch._dynamo.optimize("inductor")
def fn(a):
    unsqueeze_ = torch.ops.aten.unsqueeze_.default(a, 0)
    return unsqueeze_

args = [
      ((1, 1, 1, 12, 11, 3), (396, 396, 396, 33, 3, 1), torch.int64, "cpu")
       ]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]

with torch.no_grad():
    out = fn(*args)

# expected result: (396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 12, 11, 3])
print(args[0].stride(), args[0].shape) # incorrect result: (396, 396, 396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 1, 1, 12, 11, 3])
```
**Root cause**

1. [fake_tensor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#L140) is changed during [tracer.run](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/convert_frame.py#L311), then it will [pass incorrect inputs to inductor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/output_graph.py#L670).
2. example_inputs are changed during [propagate](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/mkldnn.py#L509)


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 


[ghstack-poisoned]

blzheng committed:
Update on "Inductor: fix incorrect result of inplace unsqueeze"



This pr aims to fix the incorrect result in the following test case.
```
torch._dynamo.optimize("inductor")
def fn(a):
    unsqueeze_ = torch.ops.aten.unsqueeze_.default(a, 0)
    return unsqueeze_

args = [
      ((1, 1, 1, 12, 11, 3), (396, 396, 396, 33, 3, 1), torch.int64, "cpu")
       ]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]

with torch.no_grad():
    out = fn(*args)

# expected result: (396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 12, 11, 3])
print(args[0].stride(), args[0].shape) # incorrect result: (396, 396, 396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 1, 1, 12, 11, 3])
```
**Root cause**

1. [fake_tensor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#L140) is changed during [tracer.run](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/convert_frame.py#L311), then it will [pass incorrect inputs to inductor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/output_graph.py#L670).
2. example_inputs are changed during [propagate](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/mkldnn.py#L509)


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 


[ghstack-poisoned]

jgong5 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -905,6 +908,26 @@ def __init__(self, *args, **kwargs):
+                    with torch.enable_grad():
+                        new_tensor = torch._prims.utils.clone_preserve_strides(self)
+                if self.grad is not None:
+                    self.grap = self.grad.clone_preserve_strides_storage()
```
I'm wondering whether this piece of code is well exercised by the tests...
typo: `self.grap` -> `self.grad`

jgong5 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -284,14 +284,17 @@ def mk_fake_tensor(make_meta_t):
         if maybe_memo is not None:
             return maybe_memo
-        out = FakeTensor(fake_mode, t, device)
+        if keep_leafness and not t.is_leaf:
+            out = FakeTensor(fake_mode, t, device).clone()
```
@ezyang Is there a good way of preserving `is_leaf` from `t` when we construct a `FakeTensor` around it?

ezyang replied:
You can look at how MetaConverter does it very carefully, but in general it's pretty difficult to do using the preexisting API surface in public PyTorch.

blzheng committed:
Update on "Inductor: fix incorrect result of inplace unsqueeze"



This pr aims to fix the incorrect result in the following test case.
```
torch._dynamo.optimize("inductor")
def fn(a):
    unsqueeze_ = torch.ops.aten.unsqueeze_.default(a, 0)
    return unsqueeze_

args = [
      ((1, 1, 1, 12, 11, 3), (396, 396, 396, 33, 3, 1), torch.int64, "cpu")
       ]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]

with torch.no_grad():
    out = fn(*args)

# expected result: (396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 12, 11, 3])
print(args[0].stride(), args[0].shape) # incorrect result: (396, 396, 396, 396, 396, 396, 33, 3, 1) torch.Size([1, 1, 1, 1, 1, 1, 12, 11, 3])
```
**Root cause**

1. [fake_tensor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/builder.py#L140) is changed during [tracer.run](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/convert_frame.py#L311), then it will [pass incorrect inputs to inductor](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/output_graph.py#L670).
2. example_inputs are changed during [propagate](https://github.com/pytorch/pytorch/blob/master/torch/_inductor/mkldnn.py#L509)


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire 


[ghstack-poisoned]
